src/token.ts

```diff
@@ -1,8 +1,393 @@
+import { regex } from './regex';
+
 export interface Token {
     type: string,
     is_title?: boolean,
     text?: string,
     scene_number?: string,
     dual?: string,
     depth?: number
+
+    addTo(tokens: Token[]): Token[]
+}
+
+export interface Block {
+    tokens: Token[],
+
+    addTo(tokens: Token[]): Token[]
+}
+
+export class TitlePageBlock implements Block {
+    readonly tokens: TitlePageToken[] = []
+
+    constructor(line: string) {
+        const match = line.replace(regex.title_page, '\n$1').split(regex.splitter).reverse();
+        this.tokens = match.reduce(
+            (previous, item) => new TitlePageToken(item).addTo(previous)
+        , [])
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, ...this.tokens]
+    }
+
+    static matchedBy(line: string): boolean {
+        return regex.title_page.test(line)
+    }
+}
+
+export class TitlePageToken implements Token {
+    readonly type: string
+    readonly is_title = true
+    readonly text: string
+
+    constructor(item: string) {
+        const pair = item.replace(regex.cleaner, '').split(/\:\n*/);
+        this.type = pair[0].trim().toLowerCase().replace(' ', '_')
+        this.text = pair[1].trim()
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        return [...tokens, this]
+    }
+}
+
+export class SceneHeadingToken implements Token {
+    readonly type = 'scene_heading'
+    readonly text: string
+    readonly scene_number: string
+    readonly too_short: boolean
+
+    constructor(line: string) {
+        const match = line.match(regex.scene_heading)
+        this.text = match[1] || match[2];
+        this.too_short = isTooShort(this.text)
+
+        const meta: RegExpMatchArray = this.text.match(regex.scene_number);
+        if (meta) {
+            this.scene_number = meta[2];
+            this.text = this.text.replace(regex.scene_number, '');
+        }
+    }
+
+    addTo(tokens: Token[]): Token[] {
+        if (this.too_short) {
+            return tokens;
+        }
```

Comment:
Currently this method when presented with something like:

```
. [[One space after the '.']]

STEEL walks in.
```

Skips the line with the period and:

```
. [[One space after the '.']]
STEEL walks in.
```

Skips both lines. 

Like with the forced character action, a line with a period and space should probably fall to action and not be skipped. I'm not sure why this was originally added into fountain.js but it's clear I didn't really think about it when I converted it to Typescript, lol.

`isTooShort` may not be a necessary here, but I think some of the regex may need to be modified since it sees `. ` as a scene heading.

That being said, this could probably be saved for another commit and to push forward with this in the PR for now.